### PR TITLE
[Merged by Bors] - Entities and components section for ECS chapter

### PR DESCRIPTION
**Pull request by alice-i-cecile**
Ported from #182.

# Status

- [x] revisit and revise
- [x] remove all use of direct world APIs
- [x] distinguish between `Entity` type and entity concept

### CODE REVIEW

File: content/learn/book/ecs/entities-components/_index.md [hunk 1]
```diff
@@ -6,6 +6,307 @@ page_template = "book-section.html"
...
+Once an entity is spawned, you can use [`Commands`] to add and remove components from them dynamically.
+
+```rust
+
+#[derive(Component)]
+struct InCombat;
+
+// This query returns the `Entity` identifier of all entities
+// that have the `Combatant` component but do not yet have the `InCombat` component
+fn start_combat_system(query: Query<Entity, (With<Combatant>, Without<InCombat>>), mut commands: Commands){
```
Comment (doup) [inline content/learn/book/ecs/entities-components/_index.md:137]:
```suggestion
fn start_combat_system(query: Query<Entity, (With<Combatant>, Without<InCombat>)>, mut commands: Commands){
```

File: content/learn/book/ecs/entities-components/_index.md [hunk 2]
```diff
@@ -6,6 +6,307 @@ page_template = "book-section.html"
...
+    for entity in query.iter(){
+        // The component will be inserted at the end of the current stage
+        commands.entity(entity).insert(InCombat);
+    }
+}
+
+// Now to undo our hard work
+fn end_combat_system(query: Query<Entity, (With<Combatant>, With<InCombat>>), mut commands: Commands){
```
Comment (doup) [inline content/learn/book/ecs/entities-components/_index.md:145]:
```suggestion
fn end_combat_system(query: Query<Entity, (With<Combatant>, With<InCombat>)>, mut commands: Commands){
```

**Review by doup — COMMENTED** [review]
I think, I saw a couple of typos on the example queries.

File: content/learn/book/ecs/entities-components/_index.md [hunk 3]
```diff
@@ -6,6 +6,307 @@ page_template = "book-section.html"
...
+Entities are entirely bare when they're spawned: they contain no data other than their unique [`Entity`] identifier.
+This of course is not very useful, so let's discuss how we can add and remove components to them which store data and enable behavior through systems.
+
+[`Entity`]: https://docs.rs/bevy/latest/bevy/ecs/entity/struct.Entity.html
+
+### Defining components
+
+To define a component type, we simply implement the [`Component`] [trait](https://doc.rust-lang.org/book/ch10-02-traits.html) for a Rust type of our choice.
+You will almost always want to use the `#[derive(Component)]` [macro](https://doc.rust-lang.org/reference/attributes/derive.html) to do this for you; which quickly and reliably generates the correct trait implementation.
+Any underlying component data must be `Send + Sync + 'static` (enforced by the [trait bounds](https://doc.rust-lang.org/book/ch10-02-traits.html#trait-bound-syntax) on [`Component`]).
```
Comment (Nilirad) [inline content/learn/book/ecs/entities-components/_index.md:54]: Soon, this won't be true anymore.

- bevyengine/bevy#4680

This is especially true if the book gets published after 0.8, for which the PR is probably going to be already merged in.

**Review by Nilirad — APPROVED** [review]
This section is quite solid. I have a couple of perplexities about the terminology around entities but those are just personal ones.

File: content/learn/book/ecs/entities-components/_index.md (hunk 3 above)
Comment (alice-i-cecile) [reply content/learn/book/ecs/entities-components/_index.md:54]: Yeah, we should cut this bit from the book intro and have a section about the trait bounds on the Component docs IMO.

File: content/learn/book/ecs/entities-components/_index.md [hunk 4]
```diff
@@ -6,6 +6,307 @@ page_template = "book-section.html"
...
+    // Selecting and then despawning the just-spawned second entity
+    commands.entity(my_entity).despawn();
+}
+```
+
+[`Commands`]: https://docs.rs/bevy/latest/bevy/ecs/system/struct.Commands.html
+
+## Working with components
+
+Entities are entirely bare when they're spawned: they contain no data other than their unique [`Entity`] identifier.
```
[the remark this answers is not in the document]
Comment (alice-i-cecile) [reply content/learn/book/ecs/entities-components/_index.md]: I agree, but not something I can do here.

**Comment by alice-i-cecile** [comment]
bors retry

**Pull request closed without merging**